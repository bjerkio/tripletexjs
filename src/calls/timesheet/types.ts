import { DefaultTripletexInputs } from '../../types';

export interface ListTimesheetEntriesInput extends DefaultTripletexInputs {
  /**
   * List of IDs
   */
  id?: string[];

  /**
   * List of IDs
   */
  employeeId?: string[];

  /**
   * List of IDs
   */
  projectId?: string[];

  /**
   * List of IDs
   */
  activityId?: string[];

  /**
   * From and including
   */
  dateFrom: Date;

  /**
   * To and including
   */
  dateTo: Date;

  /**
   * Containing
   */
  comment?: string;
}

// export interface TimeClockInput {
//   date: Date;
//   timeStart: Date;
//   timeStop: Date;
// }

export interface TimesheetEntryInput {
  projectId?: number;
  employeeId: number;
  activityId: number;
  date: Date;
  hours: number;
  comment?: string;
  // TODO: Add timeClocks
  // timeClocks:
}

export interface UpdateTimesheetEntryInput extends TimesheetEntryInput {
  id: number;
}

export interface GetMonthByMonthNumberInput extends DefaultTripletexInputs {
  /**
   * List of IDs
   */
  employeeIds: string[];

  monthYear: Date;
}
