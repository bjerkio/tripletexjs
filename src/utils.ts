import * as rt from 'runtypes';

export function pickFromObject<
  T extends Record<string, unknown>,
  K extends keyof T,
>(subject: T, ...keys: K[]): Pick<T, K> {
  const pairs = keys
    .map(key => [key, subject[key]])
    .filter(([, val]) => val !== undefined)
    .map(([key, val]) => [key, val]);
  return Object.fromEntries(pairs);
}

export function withRuntype<T>(validator: rt.Runtype<T>) {
  return (data: unknown) => {
    return validator.check(data);
  };
}

export function multipleValuesEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      from: rt.Number,
      count: rt.Number,
      versionDigest: rt.String.nullable().optional(),
      values: rt.Array(type).asReadonly(),
    })
    .asReadonly();
}

export function singleValueEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      value: type,
    })
    .asReadonly();
}
