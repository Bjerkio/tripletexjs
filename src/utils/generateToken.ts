import { OpenAPI } from '../generated/core/OpenAPI';
import { TokenSessionService } from '../generated/services/TokenSessionService';
import { invariant } from './invariant';

/**
 * generateToken helps creating a token. It sets token for all requests.
 * 
 * @param consumerToken Token of the API consumer
 * @param employeeToken The employees token
 * @param expirationDate Expiration date for the combined token
 * @example
 * ```typescript
 * const expDate = new Date();
 * expDate.setDate(expDate.getDate() + 1);
 * await generateToken({
 *   consumerToken: '000-000-000',
 *   employeeToken: '100-100-100',
 *   expirationDate: expDate,
 * });
 * ```
 */
export async function generateToken({
  expirationDate,
  ...tokenSessionArgs
}: {
  consumerToken: string;
  employeeToken: string;
  expirationDate: Date;
}) {
  if (OpenAPI.TOKEN) {
    return false;
  }

  const expDate = expirationDate.toISOString().split('T')[0];

  const session = await TokenSessionService.tokenSessionCreateCreate({
    expirationDate: expDate,
    ...tokenSessionArgs,
  });

  invariant(session.value.token, 'Not able to get token');

  OpenAPI.TOKEN = session.value;
}

/**
 * 0 or blank means the company of the employee.
 * Any other value means accountant clients. Use /company/>withLoginAccess to get a list of those.
 * @param username defaults to '0'
 */
export async function setUsername(username: string | (() => Promise<string>)) {
  OpenAPI.USERNAME = username;
}