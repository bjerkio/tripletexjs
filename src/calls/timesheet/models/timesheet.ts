import * as rt from 'runtypes';
import { multipleValuesEnvelope, singleValueEnvelope } from '../../../utils';

const timesheetRt = rt.Record({
  id: rt.Number,
  version: rt.Number,
  project: rt
    .Record({
      id: rt.Number,
    })
    .nullable()
    .optional(),
  activity: rt.Record({
    id: rt.Number,
  }),
  /**
   * TODO: Change to Date when Transform is landed, read more: https://github.com/pelotom/runtypes/pull/191
   */
  date: rt.String,
  hours: rt.Number,
  chargeableHours: rt.Number.optional(),
  employee: rt.Record({
    id: rt.Number,
  }),
  // timeClocks: rt.Array(rt.String),
  comment: rt.String.optional(),
  locked: rt.Boolean,
  chargeable: rt.Boolean.optional(),
  hourlyRate: rt.Number,
  hourlyCost: rt.Number,
  hourlyCostPercentage: rt.Number,
});

export type Timesheet = rt.Static<typeof timesheetRt>;

export const listTimesheetResponseRt = multipleValuesEnvelope(timesheetRt);
export const addTimesheetResponseRt = singleValueEnvelope(timesheetRt);
