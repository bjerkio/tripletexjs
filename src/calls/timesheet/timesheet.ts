import omitEmpty from 'omit-empty';
import {
  formatDate,
  formatMonthYear,
  serializeQuery,
  withRuntype,
} from '../../utils';
import { TripletexBase } from '../base';
import {
  addTimesheetResponseRt,
  listTimesheetResponseRt,
} from './models/timesheet';
import { getTimesheetMonthResponseRt } from './models/timesheet-month';
import {
  GetMonthByMonthNumberInput,
  ListTimesheetEntriesInput,
  TimesheetEntryInput,
  UpdateTimesheetEntryInput,
} from './types';
export * from './types';
export * from './models/timesheet';
export * from './models/timesheet-month';

export class TripletexTimesheet extends TripletexBase {
  listEntries(input: ListTimesheetEntriesInput) {
    const call = this.authenticatedCall() //
      .args<{
        input: ListTimesheetEntriesInput;
      }>()
      .path('/v2/timesheet/entry')
      .query(args => serializeQuery(args.input))
      .method('get')
      .parseJson(withRuntype(listTimesheetResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }

  updateEntries(entries: UpdateTimesheetEntryInput[]) {
    const call = this.authenticatedCall() //
      .path('/v2/timesheet/entry/list')
      .body(() => {
        return entries.map(input => {
          const entry: Record<string, any> = {
            id: input.id,
            employee: {
              id: input.employeeId,
            },
            activity: {
              id: input.activityId,
            },
            date: formatDate(input.date),
            hours: input.hours,
            comment: input.comment,
          };

          if (input.projectId) {
            entry.project = {
              id: input.projectId,
            };
          }

          return omitEmpty(entry) as Record<string, any>;
        });
      })
      .method('put')
      .parseJson(withRuntype(listTimesheetResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ sessionToken }));
  }

  addEntry(input: TimesheetEntryInput) {
    type Args = { input: TimesheetEntryInput };
    const call = this.authenticatedCall()
      .args<Args>()
      .path('/v2/timesheet/entry')
      .body(({ input }: Args) => {
        const entry: Record<string, any> = {
          employee: {
            id: input.employeeId,
          },
          activity: {
            id: input.activityId,
          },
          date: formatDate(input.date),
          hours: input.hours,
          comment: input.comment,
        };

        if (input.projectId) {
          entry.project = {
            id: input.projectId,
          };
        }

        return omitEmpty(entry) as Record<string, any>;
      })
      .method('post')
      .parseJson(withRuntype(addTimesheetResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }

  addEntries(entries: TimesheetEntryInput[]) {
    const call = this.authenticatedCall() //
      .path('/v2/timesheet/entry/list')
      .body(() => {
        return entries.map(input => {
          const entry: Record<string, any> = {
            employee: {
              id: input.employeeId,
            },
            activity: {
              id: input.activityId,
            },
            date: formatDate(input.date),
            hours: input.hours,
            comment: input.comment,
          };

          if (input.projectId) {
            entry.project = {
              id: input.projectId,
            };
          }

          return omitEmpty(entry) as Record<string, any>;
        });
      })
      .method('post')
      .parseJson(withRuntype(listTimesheetResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ sessionToken }));
  }

  getMonthByMonthNumber(input: GetMonthByMonthNumberInput) {
    const call = this.authenticatedCall() //
      .args<{
        input: GetMonthByMonthNumberInput;
      }>()
      .path('/v2/timesheet/month/byMonthNumber')
      .query(({ input: { monthYear, ...q } }) => ({
        ...serializeQuery(q),
        monthYear: formatMonthYear(monthYear),
      }))
      .method('get')
      .parseJson(withRuntype(getTimesheetMonthResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }
}
