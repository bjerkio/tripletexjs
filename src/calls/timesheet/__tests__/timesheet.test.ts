import { endOfMonth, startOfMonth } from 'date-fns';
import nock from 'nock';
import { invariant } from 'ts-invariant';
import { formatDate, parseRuntypeValidationError } from '../../../utils';
import { TripletexTimesheet } from '../timesheet';

const baseUrl = process.env.BASE_URL || 'https://api.tripletex.io/';
const sessionToken = process.env.SESSION_TOKEN || 'a-session-token';

describe('activity class', () => {
  let client: TripletexTimesheet;

  beforeEach(() => {
    client = new TripletexTimesheet({ baseUrl }, sessionToken);
  });

  it('should list timesheet entries', async () => {
    const dateFrom = startOfMonth(new Date('2022-01-01'));
    const dateTo = endOfMonth(new Date('2022-01-01'));
    nock(baseUrl, { encodedQueryParams: true })
      .get('/v2/timesheet/entry')
      .query({ dateFrom: formatDate(dateFrom), dateTo: formatDate(dateTo) })
      .reply(200, {
        fullResultSize: 6,
        from: 0,
        count: 6,
        versionDigest: "'If-None-Match' header not specified",
        values: [
          {
            id: 1211627,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/entry/1211627',
            project: null,
            activity: {
              id: 45239,
              url: 'api.tripletex.io/v2/activity/45239',
            },
            date: '2022-01-01',
            hours: 4,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: '',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
          {
            id: 1211628,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/entry/1211628',
            project: null,
            activity: {
              id: 45239,
              url: 'api.tripletex.io/v2/activity/45239',
            },
            date: '2022-01-02',
            hours: 4,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: '',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
          {
            id: 1211629,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/entry/1211629',
            project: null,
            activity: {
              id: 45239,
              url: 'api.tripletex.io/v2/activity/45239',
            },
            date: '2022-01-04',
            hours: 7,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: '',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
          {
            id: 1211630,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/entry/1211630',
            project: null,
            activity: {
              id: 45239,
              url: 'api.tripletex.io/v2/activity/45239',
            },
            date: '2022-01-05',
            hours: 4,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: '',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
          {
            id: 1211631,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/entry/1211631',
            project: null,
            activity: {
              id: 45239,
              url: 'api.tripletex.io/v2/activity/45239',
            },
            date: '2022-01-06',
            hours: 2,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: '',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
          {
            id: 1211632,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/entry/1211632',
            project: null,
            activity: {
              id: 45240,
              url: 'api.tripletex.io/v2/activity/45240',
            },
            date: '2022-01-03',
            hours: 4,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: '',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
        ],
        sumAllHours: 25,
      });
    const entries = await client.listEntries({
      dateFrom,
      dateTo,
    });

    parseRuntypeValidationError(entries.error);
    invariant(entries.success);
    expect(entries.body.values).toMatchSnapshot();
  });

  it('should update timesheet entries', async () => {
    nock(baseUrl, { encodedQueryParams: true })
      .put('/v2/timesheet/entry/list', [
        {
          id: 1211634,
          employee: { id: 1734852 },
          activity: { id: 45239 },
          date: '2022-01-31',
          hours: 15,
          comment: 'This is from bjerkio/tripletexjs',
        },
      ])
      .reply(200, {
        fullResultSize: 0,
        from: 0,
        count: 1,
        versionDigest: null,
        values: [
          {
            id: 1211634,
            version: 1,
            url: 'api.tripletex.io/v2/timesheet/entry/1211634',
            project: null,
            activity: {
              id: 45239,
              url: 'api.tripletex.io/v2/activity/45239',
            },
            date: '2022-01-31',
            hours: 15,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: 'This is from bjerkio/tripletexjs',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
        ],
      });
    const result = await client.updateEntries([
      {
        id: 1211634,
        employeeId: 1734852,
        activityId: 45239,
        date: new Date('2022-01-31'),
        hours: 15,
        comment: 'This is from bjerkio/tripletexjs',
      },
    ]);
    parseRuntypeValidationError(result.error);
    invariant(result.success);
    expect(result).toMatchSnapshot();
  });

  it('should add timesheet entries', async () => {
    nock(baseUrl, { encodedQueryParams: true })
      .post('/v2/timesheet/entry/list', [
        {
          employee: { id: 1734852 },
          activity: { id: 45239 },
          date: '2022-01-31',
          hours: 20,
          comment: 'This is from bjerkio/tripletexjs',
        },
      ])
      .reply(201, {
        fullResultSize: 0,
        from: 0,
        count: 1,
        versionDigest: null,
        values: [
          {
            id: 1211634,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/entry/1211634',
            project: null,
            activity: {
              id: 45239,
              url: 'api.tripletex.io/v2/activity/45239',
            },
            date: '2022-01-31',
            hours: 20,
            chargeableHours: 0,
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timeClocks: [],
            comment: 'This is from bjerkio/tripletexjs',
            locked: false,
            chargeable: false,
            invoice: null,
            hourlyRate: 0,
            hourlyCost: 0,
            hourlyCostPercentage: 0,
          },
        ],
      });
    const result = await client.addEntries([
      {
        employeeId: 1734852,
        activityId: 45239,
        date: new Date('2022-01-31'),
        hours: 20,
        comment: 'This is from bjerkio/tripletexjs',
      },
    ]);
    parseRuntypeValidationError(result.error);
    invariant(result.success);
    expect(result).toMatchSnapshot();
  });

  it('should get status on timesheet month', async () => {
    nock(baseUrl, { encodedQueryParams: true })
      .get('/v2/timesheet/month/byMonthNumber')
      .query({ employeeIds: '1734852', monthYear: '2022-03' })
      .reply(200, {
        fullResultSize: 0,
        from: 0,
        count: 0,
        versionDigest: null,
        values: [
          {
            id: 83,
            version: 0,
            url: 'api.tripletex.io/v2/timesheet/month/83',
            employee: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            timesheetEntries: [
              {
                id: 1211635,
                url: 'api.tripletex.io/v2/timesheet/entry/1211635',
              },
            ],
            approvedDate: null,
            completed: false,
            approvedBy: null,
            approved: false,
            approvedUntilDate: null,
            monthYear: '2022-03',
            hoursPayout: 0,
            vacationPayout: 0,
            hourSummary: {
              sumHours: 31.41,
              hoursWithPay: 31.41,
              hourlyWageHoursWithPay: 0,
              standardTime: 172.5,
              nonChargeableHours: 31.41,
              chargeableHours: 0,
              nonChargeableHoursWithPay: 31.41,
              budgetChargeableHours: 0,
            },
            flexSummary: {
              incomingHourBalance: -226.64,
              outgoingHourBalance: -367.73,
              change: -141.09,
            },
            vacationSummary: {
              incomingVacationBalance: 24.466666666666665,
              outgoingVacationBalance: 24.466666666666665,
              vacationTakenInPeriod: 0,
              vacationTakenThisYear: 0.5333333333333333,
            },
          },
        ],
      });
      
    const entries = await client.getMonthByMonthNumber({
      employeeIds: ['1734852'],
      monthYear: new Date('2022-03-01'),
    });

    parseRuntypeValidationError(entries.error);
    invariant(entries.success);
    expect(entries.body.values).toMatchSnapshot();
  });
});
