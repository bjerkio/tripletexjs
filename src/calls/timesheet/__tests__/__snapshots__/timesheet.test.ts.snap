// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity class should add timesheet entries 1`] = `
Object {
  "body": Object {
    "count": 1,
    "from": 0,
    "fullResultSize": 0,
    "values": Array [
      Object {
        "activity": Object {
          "id": 45239,
          "url": "api.tripletex.io/v2/activity/45239",
        },
        "chargeable": false,
        "chargeableHours": 0,
        "comment": "This is from bjerkio/tripletexjs",
        "date": "2022-01-31",
        "employee": Object {
          "id": 1734852,
          "url": "api.tripletex.io/v2/employee/1734852",
        },
        "hourlyCost": 0,
        "hourlyCostPercentage": 0,
        "hourlyRate": 0,
        "hours": 20,
        "id": 1211634,
        "invoice": null,
        "locked": false,
        "project": null,
        "timeClocks": Array [],
        "url": "api.tripletex.io/v2/timesheet/entry/1211634",
        "version": 0,
      },
    ],
    "versionDigest": null,
  },
  "error": undefined,
  "success": true,
}
`;

exports[`activity class should get status on timesheet month 1`] = `
Array [
  Object {
    "approved": false,
    "approvedBy": null,
    "approvedDate": null,
    "approvedUntilDate": null,
    "completed": false,
    "employee": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "flexSummary": Object {
      "change": -141.09,
      "incomingHourBalance": -226.64,
      "outgoingHourBalance": -367.73,
    },
    "hourSummary": Object {
      "budgetChargeableHours": 0,
      "chargeableHours": 0,
      "hourlyWageHoursWithPay": 0,
      "hoursWithPay": 31.41,
      "nonChargeableHours": 31.41,
      "nonChargeableHoursWithPay": 31.41,
      "standardTime": 172.5,
      "sumHours": 31.41,
    },
    "hoursPayout": 0,
    "id": 83,
    "monthYear": "2022-03",
    "timesheetEntries": Array [
      Object {
        "id": 1211635,
        "url": "api.tripletex.io/v2/timesheet/entry/1211635",
      },
    ],
    "url": "api.tripletex.io/v2/timesheet/month/83",
    "vacationPayout": 0,
    "vacationSummary": Object {
      "incomingVacationBalance": 24.466666666666665,
      "outgoingVacationBalance": 24.466666666666665,
      "vacationTakenInPeriod": 0,
      "vacationTakenThisYear": 0.5333333333333333,
    },
    "version": 0,
  },
]
`;

exports[`activity class should list timesheet entries 1`] = `
Array [
  Object {
    "activity": Object {
      "id": 45239,
      "url": "api.tripletex.io/v2/activity/45239",
    },
    "chargeable": false,
    "chargeableHours": 0,
    "comment": "",
    "date": "2022-01-01",
    "employee": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "hourlyCost": 0,
    "hourlyCostPercentage": 0,
    "hourlyRate": 0,
    "hours": 4,
    "id": 1211627,
    "invoice": null,
    "locked": false,
    "project": null,
    "timeClocks": Array [],
    "url": "api.tripletex.io/v2/timesheet/entry/1211627",
    "version": 0,
  },
  Object {
    "activity": Object {
      "id": 45239,
      "url": "api.tripletex.io/v2/activity/45239",
    },
    "chargeable": false,
    "chargeableHours": 0,
    "comment": "",
    "date": "2022-01-02",
    "employee": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "hourlyCost": 0,
    "hourlyCostPercentage": 0,
    "hourlyRate": 0,
    "hours": 4,
    "id": 1211628,
    "invoice": null,
    "locked": false,
    "project": null,
    "timeClocks": Array [],
    "url": "api.tripletex.io/v2/timesheet/entry/1211628",
    "version": 0,
  },
  Object {
    "activity": Object {
      "id": 45239,
      "url": "api.tripletex.io/v2/activity/45239",
    },
    "chargeable": false,
    "chargeableHours": 0,
    "comment": "",
    "date": "2022-01-04",
    "employee": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "hourlyCost": 0,
    "hourlyCostPercentage": 0,
    "hourlyRate": 0,
    "hours": 7,
    "id": 1211629,
    "invoice": null,
    "locked": false,
    "project": null,
    "timeClocks": Array [],
    "url": "api.tripletex.io/v2/timesheet/entry/1211629",
    "version": 0,
  },
  Object {
    "activity": Object {
      "id": 45239,
      "url": "api.tripletex.io/v2/activity/45239",
    },
    "chargeable": false,
    "chargeableHours": 0,
    "comment": "",
    "date": "2022-01-05",
    "employee": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "hourlyCost": 0,
    "hourlyCostPercentage": 0,
    "hourlyRate": 0,
    "hours": 4,
    "id": 1211630,
    "invoice": null,
    "locked": false,
    "project": null,
    "timeClocks": Array [],
    "url": "api.tripletex.io/v2/timesheet/entry/1211630",
    "version": 0,
  },
  Object {
    "activity": Object {
      "id": 45239,
      "url": "api.tripletex.io/v2/activity/45239",
    },
    "chargeable": false,
    "chargeableHours": 0,
    "comment": "",
    "date": "2022-01-06",
    "employee": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "hourlyCost": 0,
    "hourlyCostPercentage": 0,
    "hourlyRate": 0,
    "hours": 2,
    "id": 1211631,
    "invoice": null,
    "locked": false,
    "project": null,
    "timeClocks": Array [],
    "url": "api.tripletex.io/v2/timesheet/entry/1211631",
    "version": 0,
  },
  Object {
    "activity": Object {
      "id": 45240,
      "url": "api.tripletex.io/v2/activity/45240",
    },
    "chargeable": false,
    "chargeableHours": 0,
    "comment": "",
    "date": "2022-01-03",
    "employee": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "hourlyCost": 0,
    "hourlyCostPercentage": 0,
    "hourlyRate": 0,
    "hours": 4,
    "id": 1211632,
    "invoice": null,
    "locked": false,
    "project": null,
    "timeClocks": Array [],
    "url": "api.tripletex.io/v2/timesheet/entry/1211632",
    "version": 0,
  },
]
`;

exports[`activity class should update timesheet entries 1`] = `
Object {
  "body": Object {
    "count": 1,
    "from": 0,
    "fullResultSize": 0,
    "values": Array [
      Object {
        "activity": Object {
          "id": 45239,
          "url": "api.tripletex.io/v2/activity/45239",
        },
        "chargeable": false,
        "chargeableHours": 0,
        "comment": "This is from bjerkio/tripletexjs",
        "date": "2022-01-31",
        "employee": Object {
          "id": 1734852,
          "url": "api.tripletex.io/v2/employee/1734852",
        },
        "hourlyCost": 0,
        "hourlyCostPercentage": 0,
        "hourlyRate": 0,
        "hours": 15,
        "id": 1211634,
        "invoice": null,
        "locked": false,
        "project": null,
        "timeClocks": Array [],
        "url": "api.tripletex.io/v2/timesheet/entry/1211634",
        "version": 1,
      },
    ],
    "versionDigest": null,
  },
  "error": undefined,
  "success": true,
}
`;
